### ISSUE TITLE
'minishift start' fails when behind https proxy

### ISSUE THREAD

**Issue by klauern** [minishift/minishift#1534]
I think this **might** be related to #391 but I'm not sure.  In any event, i know that I am behind a corporate proxy server, so the HTTPS certificates that will come back aren't going to match anything OOTB:

```
$  minishift start                          
-- Checking if KVM driver is installed ...    
   Driver is available at /usr/bin/docker-machine-driver-kvm ...                            
   Checking driver binary is executable ... OK                                              
-- Checking if Libvirt is installed ... OK    
-- Checking if Libvirt default network is present and active ... OK                         
-- Starting local OpenShift cluster using 'kvm' hypervisor ...                              
-- Starting Minishift VM ............... OK   
-- Checking for IP address ... OK             
-- Checking if external host is reachable from the Minishift VM ...                         
   Pinging 8.8.8.8 ... FAIL                   
   VM is unable to ping external host         
-- Checking HTTP connectivity from the VM ...                                               
   Retrieving http://minishift.io/index.html ... FAIL                                       
   VM cannot connect to external URL with HTTP                                              
-- Checking if persistent storage volume is mounted ... OK                                  
-- Checking available disk space ... 0% used OK                                             
-- OpenShift cluster will be configured with ...                                            
   Version: v3.6.0                            
-- Checking `oc` support for startup flags ...                                              
   host-volumes-dir ... OK                    
   http-proxy ... OK                          
   https-proxy ... OK                         
   no-proxy ... OK                            
   routing-suffix ... OK                      
   host-config-dir ... OK                     
   host-data-dir ... OK                       
   host-pv-dir ... OK                         
Starting OpenShift using openshift/origin:v3.6.0 ...                                        
Pulling image openshift/origin:v3.6.0         
-- Checking OpenShift client ... OK           
-- Checking Docker client ... OK              
-- Checking Docker version ... OK             
-- Checking for existing OpenShift container ... OK                                         
-- Checking for openshift/origin:v3.6.0 image ...                                           
   Pulling image openshift/origin:v3.6.0      
FAIL                                          
   Error: error pulling Docker image openshift/origin:v3.6.0                                
   Caused By:                                 
     Error: Error while pulling image: Get https://index.docker.io/v1/repositories/openshift/origin/images: x509: certificate signed by unknown authority                               
Error during 'cluster up' execution: Error starting the cluster.                       
```

**Comment by praveenkumar** [comment]
@karmab Can you do `minishift ssh` and then paste output of `docker info` here? As per your logs  (below output) I think proxy is not even set inside the VM.

```
-- Checking if external host is reachable from the Minishift VM ...                         
   Pinging 8.8.8.8 ... FAIL                   
   VM is unable to ping external host         
-- Checking HTTP connectivity from the VM ...                                               
   Retrieving http://minishift.io/index.html ... FAIL 
```

**Comment by klauern** [comment]
It does appear to be set.  I have to omit the details because of security reasons, but here's what I have

```
docker@minishift:~$ docker info               
Containers: 0                                 
 Running: 0                                   
 Paused: 0                                    
 Stopped: 0                                   
Images: 0                                     
Server Version: 1.12.3                        
Storage Driver: aufs                          
 Root Dir: /mnt/sda1/var/lib/docker/aufs      
 Backing Filesystem: extfs                    
 Dirs: 0                                      
 Dirperm1 Supported: true                     
Logging Driver: json-file                     
Cgroup Driver: cgroupfs                       
Plugins:                                      
 Volume: local                                
 Network: null host bridge overlay            
Swarm: inactive                               
Runtimes: runc                                
Default Runtime: runc                         
Security Options: seccomp                     
Kernel Version: 4.4.27-boot2docker            
Operating System: Minishift Boot2Docker ISO Version: 1.0.2                                  
OSType: linux                                 
Architecture: x86_64                          
CPUs: 2                                       
Total Memory: 1.955 GiB                       
Name: minishift                               
ID: KT3O:QNRW:QVAU:P4VW:UFD2:BJ6B:MFIF:GODG:U5NI:DA3L:7EFI:YKON                             
Docker Root Dir: /mnt/sda1/var/lib/docker     
Debug Mode (client): false                    
Debug Mode (server): true                     
 File Descriptors: 14                         
 Goroutines: 23                               
 System Time: 2017-10-10T15:07:06.975104732Z  
 EventsListeners: 0                           
Http Proxy: proxy.work-thing.com:80   
Https Proxy: http://proxy.work-thing.com:80                                         
No Proxy: localhost,127.0.0.1,172.30.1.1,172.17.0.*,.localhost,localhost*,127.0.0.1,.dockerhost,0.0.0.0,.workthing.com*,.work-thing.com*,.work.thing*                                        
Registry: https://index.docker.io/v1/         
Labels:                                       
 provider=kvm                                 
Insecure Registries:                          
 172.30.0.0/16                                
 127.0.0.0/8                                  
```

Could it be the difference between having the `http` and not?  Which is the "right" choice?

**Comment by klauern** [comment]
@gbraad Correct.  It's an internal CA that resides on an intranet.  All external requests go through the proxy, yadda yadda *mumble something about security*, and come back re-signed with our internal CA, which is probably sniffed out for nefarious stuff.

**Comment by gbraad** [comment]
This might be the problem, as the CA might not get recognized in the VM.

**Comment by klauern** [comment]
I'm not sure that works, not sure why:

```
$  openssl s_client -showcerts -connect proxy.work-thing.com:80                   
CONNECTED(00000003)                           
140449783082944:error:1408F10B:SSL routines:ssl3_get_record:wrong version number:ssl/record/ssl3_record.c:252:                                                                          
---                                           
no peer certificate available                 
---                                           
No client certificate CA names sent           
---                                           
SSL handshake has read 5 bytes and written 176 bytes                                        
Verification: OK                              
---                                           
New, (NONE), Cipher is (NONE)                 
Secure Renegotiation IS NOT supported         
Compression: NONE                             
Expansion: NONE                               
No ALPN negotiated                            
SSL-Session:                                  
    Protocol  : TLSv1.2                       
    Cipher    : 0000                          
    Session-ID:                               
    Session-ID-ctx:                           
    Master-Key:                               
    PSK identity: None                        
    PSK identity hint: None                   
    SRP username: None                        
    Start Time: 1507653444                    
    Timeout   : 7200 (sec)                    
    Verify return code: 0 (ok)                
    Extended master secret: no                
--- 
```

I know for Docker itself, I can pull images and such from DockerHub, so I assume it's using my `/etc/ssl/cafile.pem` or some equivalent that I have imported these as trusted certificates for.

**Comment by gbraad** [comment]
Slightly off-topic: @praveenkumar we might also have to update the preflight check to use:

`CURLOPT_PROXY`, `CURLOPT_PROXYUSERPWD` and `CURLOPT_PROXYTYPE` and set like `CURLPROXY_HTTPS` => https://curl.haxx.se/libcurl/c/CURLOPT_PROXYTYPE.html

**Comment by gbraad** [comment]
when you say:

> I know for Docker itself, I can pull images

you meant your local machine, right?

**Comment by klauern** [comment]
Correct.

**Comment by klauern** [comment]
I assume that if I can `docker pull openshift/origin:v.3.6.0`, I would be able to get past this, but that wouldn't solve **why** it's not able to do it by itself.

**Comment by gbraad** [comment]
It is possible to import this image into the VM, as part of image-caching. But I am looking into a possible workaround... using an upstream/forwarding proxy, with for instance nginx, (or squid). In that case you let Minishift connect to your local proxy, which then forwards to work-thing:80

**Comment by gbraad** [comment]
@praveenkumar can you setup a proxy with a self-created CA signed certificate?

**Comment by ckolumbus** [comment]
I use [privoxy](http://www.privoxy.org/) as local forwarding proxy (instead of the unmaintained polipo and without the privacy options enabled) and it listens to the virtualbox network addresses, eg. 192.168.56.1:3128.

when I set the http_proxy / https_proxy environement variables to http://192.168.56.1:3128 BEFORE starting minishift for the first time, then it is able to pull all necessary images.

**Comment by gbraad** [comment]
Privoxy was the other one I was looking for, Thanks!

We are considering to include our own proxy, as this can also help with exposing ports.

**Comment by praveenkumar** [comment]
@ckolumbus Can you also provide which host operating system you used for privoxy experiment?

**Comment by praveenkumar** [comment]
Closing this issue, we already have follow-up issue #1628 for same.

**Comment by ckolumbus** [comment]
@praveenkumar Sorry for the (too?) late reply: I was on a corporate Win7 machine.

**Comment by darkedges** [comment]
privoxy does not work for the corporate firewall I am behind.

```
docker@minishift:~$ env | grep -i http
https_proxy=http://10.xxx.xx.xx:8118
http_proxy=http://10.xxx.xx.xx:8118
docker@minishift:~$ curl https://www.google.com.au
curl: (60) SSL certificate problem: unable to get local issuer certificate
More details here: https://curl.haxx.se/docs/sslcerts.html

curl performs SSL certificate verification by default, using a "bundle"
 of Certificate Authority (CA) public keys (CA certs). If the default
 bundle file isn't adequate, you can specify an alternate file
 using the --cacert option.
If this HTTPS server uses a certificate signed by a CA represented in
 the bundle, the certificate verification probably failed due to a
 problem with the certificate (it might be expired, or the name might
 not match the domain name in the URL).
If you'd like to turn off curl's verification of the certificate, use
 the -k (or --insecure) option.
```

Hence why we need an option to bake the CA Certificate into the service.

**Comment by nextropia** [comment]
You could try this workaround as a temporary fix:
https://stackoverflow.com/questions/21181231/server-certificate-verification-failed-cafile-etc-ssl-certs-ca-certificates-c/21181447#21181447

**Comment by krishnajinka** [comment]
Hello, I am facing similar issue , minishift fails while pulling docker image. However, when I pull it manually, it gets successfully. I am not sure how to tell minishift to use locally available image, or not sure why minishift fails to download image. 
I am new to minishift, could you please tell me how to circumvent this problem ?

Removed some data due to security restrictions... 

minishift start --vm-driver kvm --https-proxy "http://<proxy_ip>:8080/" --docker-env 'NO_PROXY="<dev_machine_ip>" HTTP-PROXY="http://<proxy_ip>:8080/" HTTPS-PROXY="http://<proxy_ip>:8080/" DOCKER_TLS_VERIFY=1 DOCKER_HOST=tcp://<dev_machine_ip>:2376 DOCKER_CERT_PATH=/Users/krisjinka/.minishift/certs'
-- Starting profile 'minishift'
-- Using proxy for the setup
-- Check if deprecated options are used ... OK
-- Checking if https://github.com is reachable ... OK
-- Checking if requested OpenShift version 'v3.11.0' is valid ... OK
-- Checking if requested OpenShift version 'v3.11.0' is supported ... OK
-- Checking if requested hypervisor 'kvm' is supported on this platform ... OK
-- Checking if KVM driver is installed ... 
   Driver is available at /usr/local/bin/docker-machine-driver-kvm ... 
   Checking driver binary is executable ... OK
-- Checking if Libvirt is installed ... OK
-- Checking if Libvirt default network is present ... OK
-- Checking if Libvirt default network is active ... OK
-- Checking the ISO URL ... OK
-- Checking if provided oc flags are supported ... OK
-- Starting the OpenShift cluster using 'kvm' hypervisor ...
-- Minishift VM will be configured with ...
   Memory:    4 GB
   vCPUs :    2
   Disk size: 20 GB
-- Starting Minishift VM ................-- Setting proxy information ... OK
 OK
-- Checking for IP address ... OK
-- Checking for nameservers ... OK
-- Checking if external host is reachable from the Minishift VM ... 
   Pinging 8.8.8.8 ... FAIL
   VM is unable to ping external host
-- Checking HTTP connectivity from the VM ... 
   Retrieving http://minishift.io/index.html ... OK
-- Checking if persistent storage volume is mounted ... OK
-- Checking available disk space ... 1% used OK
-- Writing current configuration for static assignment of IP address ... WARN
   Importing 'openshift/origin-control-plane:v3.11.0'  CACHE MISS
   Importing 'openshift/origin-docker-registry:v3.11.0'  CACHE MISS
   Importing 'openshift/origin-haproxy-router:v3.11.0'  CACHE MISS
-- OpenShift cluster will be configured with ...
   Version: v3.11.0
-- Pulling the OpenShift Container Image .Error pulling the OpenShift container image: ssh command error:
command : docker pull openshift/origin-control-plane:v3.11.0
err     : exit status 1
output  : Trying to pull repository docker.io/openshift/origin-control-plane ... 
Get https://registry-1.docker.io/v2/: x509: certificate signed by unknown authority

>>> virsh list
 Id    Name                           State
----------------------------------------------------
 9     minishift                      running
